docs/ci_parse_caa_reply.3
.\"
.\" Copyright 2020 Danny Sonnenschein <my.card.god@web.de>
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_PARSE_CAA_REPLY 3 "16 September 2020"
.SH NAME
ci_parse_caa_reply \- Parse a reply to a DNS query of type CAA
.SH SYNOPSIS
.nf
#include <ci.h>

int ci_parse_caa_reply(const unsigned char* \fIabuf\fP, int \fIalen\fP,
                         struct ci_caa_reply **\fIcaa_out\fP);
.fi
.SH DESCRIPTION
The
.BR "ci_parse_caa_reply" 
function parses the response to a query of type CAA into a
linked list (one element per sub-string) of
.IR "struct ci_caa_reply"
The parameters
.I abuf
and
.I alen
give the contents of the response.  The result is stored in allocated
memory and a pointer to it stored into the variable pointed to by
.IR caa_out .
It is the caller's responsibility to free the resulting
.IR caa_out
structure when it is no longer needed using the function
.B ci_free_data(3)
.PP
The structure 
.I ci_caa_reply(3)
contains the following fields:
.sp
.in +4n
.nf
struct ci_caa_reply {
  struct ci_caa_reply *next;
  int                    critical;
  unsigned char         *property;
  size_t                 plength; /* plength excludes null */
  unsigned char         *value;
  size_t                 length;  /* length excludes null */
};
.fi
.in
.PP
.SH RETURN VALUES
.BR "ci_parse_caa_reply"
can return any of the following values:
.TP 15
.B CI_SUCCESS
The response was successfully parsed.
.TP 15
.B CI_EBADRESP
The response was malformatted.
.TP 15
.B CI_ENODATA
The response did not contain an answer to the query.
.TP 15
.B CI_ENOMEM
Memory was exhausted.
.SH EXAMPLE
.nf
#include <arpa/inet.h>
#include <time.h>
#include <sys/time.h>
#include <netdb.h>

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>

#include "ci.h"

static void dns_callback(void *arg,
                         int status,
                         int timeouts,
                         unsigned char *abuf,
                         int alen)
  {
    struct ci_caa_reply *caa_out;
    int err;

    err = ci_parse_caa_reply (abuf, alen, &caa_out);
    if (err == CI_SUCCESS)
      {
        struct ci_caa_reply *caa_curr;
        for (caa_curr=caa_out; caa_curr; caa_curr=caa_curr->next)
          printf ("%s. CAA %i %s \\"%s\\"\\n", arg,
                                            caa_curr->critical,
                                            caa_curr->property,
                                            caa_curr->value);
      }
    else
      {
        printf ("err=%i\\n", err);
      }
    ci_free_data (caa_out);
  }

static void main_loop(ci_channel_t **channel)
  {
    int nfds, count;
    fd_set readers, writers;
    struct timeval tv, *tvp;
    while (1)
      {
        FD_ZERO (&readers);
        FD_ZERO (&writers);
        nfds = ci_fds (*channel, &readers, &writers);
        if (nfds == 0)
          break;
        tvp = ci_timeout (*channel, NULL, &tv);
        count = select (nfds, &readers, &writers, NULL, tvp);
        ci_process (*channel, &readers, &writers);
      }
  }

int main(int argc, char **argv)
  {
    const char *sversion;
    int iversion;
    int err;

    sversion = ci_version (&iversion);
    printf ("c-ci version %s\\n", sversion);

    char *domain = "wikipedia.org";
    if (argc > 1)
      domain = argv[1];

    ci_channel_t *channel;
    if ((err = ci_init (&channel)) != CI_SUCCESS)
      {
        printf ("ci_init() failed (%i)\\n", err);
        exit (EXIT_FAILURE);
      }

    ci_query (channel, domain,
                1,   /* ns_c_in */
                257, /* T_CAA */
                dns_callback, domain);

    main_loop (&channel);

    ci_destroy (channel);

    exit (EXIT_SUCCESS);
  }
.fi
.SH AVAILABILITY
This function was first introduced in c-ci version 1.17.0.
.SH SEE ALSO
.BR ci_query (3)
.BR ci_free_data (3)


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/legacy/ci_parse_caa_reply.c
/* MIT License
 *
 * Copyright (c) 2023 Brad House
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ci_private.h"
#include "ci_data.h"

int ci_parse_caa_reply(const unsigned char *abuf, int alen_int,
                         struct ci_caa_reply **caa_out)
{
  ci_status_t          status;
  size_t                 alen;
  struct ci_caa_reply *caa_head = NULL;
  struct ci_caa_reply *caa_last = NULL;
  struct ci_caa_reply *caa_curr;
  ci_dns_record_t     *dnsrec = NULL;
  size_t                 i;

  *caa_out = NULL;

  if (alen_int < 0) {
    return CI_EBADRESP;
  }

  alen = (size_t)alen_int;

  status = ci_dns_parse(abuf, alen, 0, &dnsrec);
  if (status != CI_SUCCESS) {
    goto done;
  }

  if (ci_dns_record_rr_cnt(dnsrec, CI_SECTION_ANSWER) == 0) {
    status = CI_ENODATA;
    goto done;
  }

  for (i = 0; i < ci_dns_record_rr_cnt(dnsrec, CI_SECTION_ANSWER); i++) {
    const unsigned char *ptr;
    size_t               ptr_len;
    const ci_dns_rr_t *rr =
      ci_dns_record_rr_get(dnsrec, CI_SECTION_ANSWER, i);

    if (rr == NULL) {
      /* Shouldn't be possible */
      status = CI_EBADRESP; /* LCOV_EXCL_LINE: DefensiveCoding */
      goto done;              /* LCOV_EXCL_LINE: DefensiveCoding */
    }

    /* XXX: Why do we allow Chaos class? */
    if (ci_dns_rr_get_class(rr) != CI_CLASS_IN &&
        ci_dns_rr_get_class(rr) != CI_CLASS_CHAOS) {
      continue;
    }

    /* Only looking for CAA records */
    if (ci_dns_rr_get_type(rr) != CI_REC_TYPE_CAA) {
      continue;
    }

    /* Allocate storage for this CAA answer appending it to the list */
    caa_curr = ci_malloc_data(CI_DATATYPE_CAA_REPLY);
    if (caa_curr == NULL) {
      status = CI_ENOMEM; /* LCOV_EXCL_LINE: OutOfMemory */
      goto done;            /* LCOV_EXCL_LINE: OutOfMemory */
    }

    /* Link in the record */
    if (caa_last) {
      caa_last->next = caa_curr;
    } else {
      caa_head = caa_curr;
    }
    caa_last = caa_curr;

    caa_curr->critical = ci_dns_rr_get_u8(rr, CI_RR_CAA_CRITICAL);
    caa_curr->property =
      (unsigned char *)ci_strdup(ci_dns_rr_get_str(rr, CI_RR_CAA_TAG));
    if (caa_curr->property == NULL) {
      status = CI_ENOMEM; /* LCOV_EXCL_LINE: OutOfMemory */
      break;                /* LCOV_EXCL_LINE: OutOfMemory */
    }
    /* RFC6844 says this can only be ascii, so not sure why we're recording a
     * length */
    caa_curr->plength = ci_strlen((const char *)caa_curr->property);

    ptr = ci_dns_rr_get_bin(rr, CI_RR_CAA_VALUE, &ptr_len);
    if (ptr == NULL) {
      status = CI_EBADRESP; /* LCOV_EXCL_LINE: DefensiveCoding */
      goto done;              /* LCOV_EXCL_LINE: DefensiveCoding */
    }

    /* Wants NULL termination for some reason */
    caa_curr->value = ci_malloc(ptr_len + 1);
    if (caa_curr->value == NULL) {
      status = CI_ENOMEM; /* LCOV_EXCL_LINE: OutOfMemory */
      goto done;            /* LCOV_EXCL_LINE: OutOfMemory */
    }
    memcpy(caa_curr->value, ptr, ptr_len);
    caa_curr->value[ptr_len] = 0;
    caa_curr->length         = ptr_len;
  }

done:
  /* clean up on error */
  if (status != CI_SUCCESS) {
    if (caa_head) {
      ci_free_data(caa_head);
    }
  } else {
    /* everything looks fine, return the data */
    *caa_out = caa_head;
  }
  ci_dns_record_destroy(dnsrec);
  return (int)status;
}
